File: classic150/Solution0076.cpp
#include <string>
#include <vector>
#include <climits>
using namespace std;
// 给你一个字符串 s 、一个字符串 t 。返回 s 中涵盖 t 所有字符的最小子串。如果 s 中不存在涵盖 t 所有字符的子串，则返回空字符串 "" 。

// 注意：

// 对于 t 中重复字符，我们寻找的子字符串中该字符数量必须不少于 t 中该字符数量。
// 如果 s 中存在这样的子串，我们保证它是唯一的答案。

// 示例 1：

// 输入：s = "ADOBECODEBANC", t = "ABC"
// 输出："BANC"
// 解释：最小覆盖子串 "BANC" 包含来自字符串 t 的 'A'、'B' 和 'C'。
// 示例 2：

// 输入：s = "a", t = "a"
// 输出："a"
// 解释：整个字符串 s 是最小覆盖子串。
// 示例 3:

// 输入: s = "a", t = "aa"
// 输出: ""
// 解释: t 中两个字符 'a' 均应包含在 s 的子串中，
// 因此没有符合条件的子字符串，返回空字符串。

// 提示：

// m == s.length
// n == t.length
// 1 <= m, n <= 105
// s 和 t 由英文字母组成

// 进阶：你能设计一个在 o(m+n) 时间内解决此问题的算法吗？
class Solution
{
public:
    string minWindow(string s, string t)
    {
        int n = s.size();
        int m = t.size();
        if (n < m)
            return "";
        vector<int> occur(128, 0);
        for (char c : t)
            occur[c]++;
        int minStart = 0, minLen = INT_MAX;
        int left = 0, right = 0;
        while (right < n)
        {
            while (right < n && m > 0)
            {
                if (occur[s[right++]]-- > 0)
                    m--;
            }
            while(m == 0)
            {
                if (right - left < minLen)
                {
                    minLen = right - left;
                    minStart = left;
                }
                if (occur[s[left++]]++ == 0)
                    m++;
            }
        }
        return minLen == INT_MAX ? "" : s.substr(minStart, minLen);
    }
};

File: anew/Solution0076.java
package anew;

import java.util.Map;
import java.util.HashMap;
// 给定两个字符串 s 和 t，长度分别是 m 和 n，返回 s 中的 最短窗口 子串，使得该子串包含 t 中的每一个字符（包括重复字符）。如果没有这样的子串，返回空字符串 ""。

// 测试用例保证答案唯一。
public class Solution0076 {
    public String minWindow(String s, String t) {
        Map<Character, Integer> need = new HashMap<>();
        Map<Character, Integer> window = new HashMap<>();
        for (char c : t.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }
        int minLen = Integer.MAX_VALUE;
        int start = 0;
        int left = 0, right = 0;
        int valid = 0;
        while (right < s.length()) {
            char rightC = s.charAt(right++);
            if (need.containsKey(rightC)) {
                window.put(rightC, window.getOrDefault(rightC, 0) + 1);
                if (window.get(rightC).equals(need.get(rightC))) {
                    valid++;
                }
            }
            while (valid == need.size()) {
                if (right - left < minLen) {
                    start = left;
                    minLen = right - left;
                }
                char leftC = s.charAt(left++);
                if (need.containsKey(leftC)) {
                    if (window.get(leftC).equals(need.get(leftC))) {
                        valid--;
                    }
                    window.put(leftC, window.get(leftC) - 1);
                }
            }
        }
        return minLen == Integer.MAX_VALUE ? "" : s.substring(start, start + minLen);
    }

    public String minWindow2(String s, String t) {
        int[] count = new int[128];
        for (char c : t.toCharArray()) {
            count[c]++;
        }
        int start = 0, minLen = Integer.MAX_VALUE;
        int left = 0, right = 0;
        int valid = t.length();
        while (right < s.length()) {
            char rightC = s.charAt(right++);
            if (count[rightC] > 0) {
                valid--;
            }
            count[rightC]--;
            while (valid == 0) {
                if (right - left < minLen) {
                    start = left;
                    minLen = right - left;
                }
                char leftC = s.charAt(left++);
                count[leftC]++;
                if (count[leftC] > 0) {
                    valid++;
                }
            }
        }
        return minLen == Integer.MAX_VALUE ? "" : s.substring(start, start + minLen);
    }

    public static void main(String[] args) {
        Solution0076 solution0076 = new Solution0076();
        String s = "ADOBECODEBANC";
        String t = "ABC";
        System.out.println(solution0076.minWindow(s, t));
        System.out.println(solution0076.minWindow2(s, t));
    }
}
